src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,1801 @@
+                        jsonReader.Skip();
+                    }
+                }
+            }
+            packageSpec.Name = name;
+            packageSpec.FilePath = name == null ? null : Path.GetFullPath(packageSpecPath);
+
+#pragma warning disable CS0612 // Type or member is obsolete
+            if (!wasPackOptionsSet)
+            {
+                packageSpec.Owners = [];
+                packageSpec.PackOptions = new PackOptions()
+                {
+                    PackageType = Array.Empty<PackageType>()
+                };
+                packageSpec.Tags = [];
+            }
+
+            if (isMappingsNull)
+            {
+                packageSpec.PackOptions.Mappings = null;
+            }
+#pragma warning restore CS0612 // Type or member is obsolete
+
+            packageSpec.RuntimeGraph = new RuntimeGraph(
+                runtimeDescriptions ?? Enumerable.Empty<RuntimeDescription>(),
+                compatibilityProfiles ?? Enumerable.Empty<CompatibilityProfile>());
+
+            packageSpec.Name ??= packageSpec.RestoreMetadata?.ProjectName;
+
+            // Use the project.json path if one is set, otherwise use the project path
+            packageSpec.FilePath ??= packageSpec.RestoreMetadata?.ProjectJsonPath
+                    ?? packageSpec.RestoreMetadata?.ProjectPath;
+
+            return packageSpec;
+        }
+
+        internal static void ReadCentralTransitiveDependencyGroup(
+            ref Utf8JsonStreamReader jsonReader,
+            IList<LibraryDependency> results,
+            string packageSpecPath)
+        {
+            if (jsonReader.Read() && jsonReader.TokenType == JsonTokenType.StartObject)
+            {
+                while (jsonReader.Read() && jsonReader.TokenType == JsonTokenType.PropertyName)
+                {
+                    var propertyName = jsonReader.GetString();
+                    if (string.IsNullOrEmpty(propertyName))
+                    {
+                        throw FileFormatException.Create(
+                            "Unable to resolve dependency ''.",
+                            packageSpecPath);
+                    }
+
+                    if (jsonReader.Read())
+                    {
+                        var dependencyIncludeFlagsValue = LibraryIncludeFlags.All;
+                        var dependencyExcludeFlagsValue = LibraryIncludeFlags.None;
+                        var suppressParentFlagsValue = LibraryIncludeFlagUtils.DefaultSuppressParent;
+                        string dependencyVersionValue = null;
+
+                        if (jsonReader.TokenType == JsonTokenType.String)
+                        {
+                            dependencyVersionValue = jsonReader.GetString();
+                        }
+                        else if (jsonReader.TokenType == JsonTokenType.StartObject)
+                        {
+                            while (jsonReader.Read() && jsonReader.TokenType == JsonTokenType.PropertyName)
```

Comment:
Maybe this code is a direct copy of the Newtonsoft.Json code, but I think it would be nice to reduce nesting within this fuction, and have different objects parsed in different functions. Since optimized code loses line numbers in exceptions, this helps investigate errors. Shorter functions are also easier to read & understand.

A [local function](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions) can be used, to keep this dependency parsing hidden outside the `ReadCentralTransitiveDependencyGroup` method, if that's desired.

Reply:
I split it across multiple functions. I can't use local functions with Ref Structs unfortunately.